### PR TITLE
Fixed the calculation of the combined speed scaling factor

### PR DESCRIPTION
**Pull request by MGBla**
I had to further adjust the hardware interface since the combined scaling factor was not calculated. Also added parts of the internal robot state machine from the ros1 driver since it was relevant for the way the factor is calculated. 

There is still some hacky stuff in here because the industrial_robot_interface is not released for foxy. I will create additional issues for this.

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
@MGBla wrote:
> I had to further adjust the hardware interface since the combined scaling factor was not calculated.

out of curiosity: what's the *combined scaling factor* ?

File: ur_robot_driver/src/hardware_interface.cpp [hunk 1]
```diff
@@ -426,6 +487,58 @@ return_type URPositionHardwareInterface::write()
...
+
+  // Motion is not possible if controller is either in error or in safeguard stop.
+  // TODO: Check status of robot program "external control" here as well
+  if (robot_status_resource_.in_error == TriState::TRUE ||
+      safety_status_bits_[urcl::toUnderlying(rtde::UrRtdeSafetyStatusBits::IS_SAFEGUARD_STOPPED)])
+  {
+    robot_status_resource_.motion_possible = TriState::FALSE;
+  }
+  // TODO The hardcoded 7 should be ur_dashboard_msgs::RobotMode::RUNNING exchange when msgs are built
+  else if (robot_mode_ == 7)
```
Comment (livanov93) [inline ur_robot_driver/src/hardware_interface.cpp:526]: Msgs are already built

**Review by livanov93 — COMMENTED** [review]
Generally comments are related to the infrastructure stuff. I advise to do changes step by step, this means you should expose robots status information first if you are going to need them.
Also information about which controller is running should be set within specialized state controllers which will get those information from pos or vel based trajectory controller.
 If you are going to include stuff like industrial_robot_status_interface please take them to a separate file.

File: ur_robot_driver/include/ur_robot_driver/hardware_interface.h [hunk 2]
```diff
@@ -113,9 +126,40 @@ class URPositionHardwareInterface : public hardware_interface::SystemInterface
   urcl::vector6d_t urcl_joint_efforts_;
   urcl::vector6d_t urcl_ft_sensor_measurements_;
   urcl::vector6d_t urcl_tcp_pose_;
+
+  // TODO exchange for industrial robot status interface
+  enum class TriState : int8_t
```
[the remark this answers is not in the document]
Comment (MGBla) [reply ur_robot_driver/include/ur_robot_driver/hardware_interface.h]: done

File: ur_robot_driver/include/ur_robot_driver/hardware_interface.h [hunk 3]
```diff
@@ -113,9 +126,40 @@ class URPositionHardwareInterface : public hardware_interface::SystemInterface
...
+  {
+    UNKNOWN = -1,
+    FALSE = 0,
+    TRUE = 1,
+  };
+  enum class RobotMode : int8_t
```
[the remark this answers is not in the document]
Comment (MGBla) [reply ur_robot_driver/include/ur_robot_driver/hardware_interface.h]: done

File: ur_robot_driver/include/ur_robot_driver/hardware_interface.h [hunk 4]
```diff
@@ -113,9 +126,40 @@ class URPositionHardwareInterface : public hardware_interface::SystemInterface
...
+  {
+    UNKNOWN = -1,
+    MANUAL = 1,
+    AUTO = 2,
+  };
+  typedef std::int32_t ErrorCode;
+  struct RobotStatus
```
[the remark this answers is not in the document]
Comment (MGBla) [reply ur_robot_driver/include/ur_robot_driver/hardware_interface.h]: done

File: ur_robot_driver/include/ur_robot_driver/hardware_interface.h [hunk 5]
```diff
@@ -113,9 +126,40 @@ class URPositionHardwareInterface : public hardware_interface::SystemInterface
...
+  {
+    RobotMode mode;
```
[the remark this answers is not in the document]
Comment (MGBla) [reply ur_robot_driver/include/ur_robot_driver/hardware_interface.h]: done

File: ur_robot_driver/include/ur_robot_driver/hardware_interface.h [hunk 6]
```diff
@@ -113,9 +126,40 @@ class URPositionHardwareInterface : public hardware_interface::SystemInterface
...
+    TriState in_error;
+    ErrorCode error_code;
+  };
+  RobotStatus robot_status_resource_{};
+  // END TODO
+
   bool packet_read_;
 
   uint32_t runtime_state_;
+  bool position_controller_running_;
```
[the remark this answers is not in the document]
Comment (MGBla) [reply ur_robot_driver/include/ur_robot_driver/hardware_interface.h]: Removed them for now

File: ur_robot_driver/include/ur_robot_driver/hardware_interface.h [hunk 7]
```diff
@@ -140,6 +184,9 @@ class URPositionHardwareInterface : public hardware_interface::SystemInterface
   bool non_blocking_read_;
   bool position_interface_in_use_;
 
+  PausingState pausing_state_;
```
[the remark this answers is not in the document]
Comment (MGBla) [reply ur_robot_driver/include/ur_robot_driver/hardware_interface.h:162]: Removed them for now

File: ur_robot_driver/src/hardware_interface.cpp (hunk 1 above)
Comment (MGBla) [reply ur_robot_driver/src/hardware_interface.cpp]: Exchanged for message enum

File: ur_robot_driver/include/ur_robot_driver/hardware_interface.h [hunk 8]
```diff
@@ -113,9 +126,40 @@ class URPositionHardwareInterface : public hardware_interface::SystemInterface
...
+    ErrorCode error_code;
+  };
+  RobotStatus robot_status_resource_{};
+  // END TODO
+
   bool packet_read_;
 
   uint32_t runtime_state_;
+  bool position_controller_running_;
+  bool velocity_controller_running_;
```
[the remark this answers is not in the document]
Comment (MGBla) [reply ur_robot_driver/include/ur_robot_driver/hardware_interface.h]: Removed them for now

File: ur_robot_driver/src/hardware_interface.cpp [hunk 9]
```diff
@@ -338,6 +354,15 @@ void URPositionHardwareInterface::readBitsetData(const std::unique_ptr<rtde::Dat
 
 return_type URPositionHardwareInterface::read()
 {
+  robot_status_resource_.mode = RobotMode::UNKNOWN;
```
[the remark this answers is not in the document]
Comment (MGBla) [reply ur_robot_driver/src/hardware_interface.cpp:354]: I think we should tackle the robot status in an independent issue. How should we proceed here @livanov93 ? I think there are two options: + Remove all status stuff for the moment and set the scaling factor independent of the current robot state
+ Leave it in for the time being and straighten everything out as soon as the robot status state controller is finished

**Comment by MGBla** [comment]
@gavanderhoorn Since you are already involved here: Currently the code contains a rather hacky re-implementation of the enumerations of your industrial_robot_status_interface, with which I'm not really happy. Are there plans to release them for Foxy in the near future?

**Comment by gavanderhoorn** [comment]
I don't really understand how `industrial_robot_status_interface` (and its controller) are involved in the calculation of the scaling factor.

Afaict, it's not needed.

**Comment by MGBla** [comment]
Oh yeah you're right... Seems I confused something there. Thought it was used the determine if the robot is currently in a ramp up phase when coming from a paused state, which is included in the combined scaling factor.

File: ur_robot_driver/src/hardware_interface.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (MGBla) [reply ur_robot_driver/src/hardware_interface.cpp]: Ok after clarifications I removed all the code relevant for the robot state

**Pull request merged**